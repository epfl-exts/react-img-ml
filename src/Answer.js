import React, { useContext, useEffect } from "react";
import { RoundContext } from "./Round.js";

const Answer = React.forwardRef(() => {
  const { ref } = useContext(RoundContext);
  let ctx;
  let mouseDown = false;
  let lastX;
  let lastY;

  function drawLine(x, y, lastX, lastY) {
    ctx.beginPath();
    ctx.strokeStyle = "#ffffff";
    ctx.lineWidth = 12;
    ctx.lineJoin = "round";
    ctx.moveTo(lastX, lastY);
    ctx.lineTo(x, y);
    ctx.closePath();
    ctx.stroke();

    return [x, y];
  }

  function handleMouseMove(event) {
    const rect = event.target.getBoundingClientRect();
    const x = event.clientX - rect.left;
    const y = event.clientY - rect.top;

    if (mouseDown) {
      [lastX, lastY] = drawLine(x, y, lastX, lastY);
    }
  }

  useEffect(() => {
    ctx = ref.current.getContext("2d");
    return null;
  });

  return (
    <canvas
      className="answer"
      onMouseDown={() => (mouseDown = true)}
      onMouseMove={e => handleMouseMove(e)}
      onMouseUp={() => {
        mouseDown = false;
        [lastX, lastY] = [undefined, undefined];
      }}
      height={320}
      ref={ref}
      width={320}
    />
  );
});

export default Answer;
