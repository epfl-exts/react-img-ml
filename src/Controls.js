import React, { useContext, useEffect, useState } from "react";
import Answer from "./Answer.js";
import { GameContext } from "./App.js";
import { predict } from "./helpers.js";

const ControlsContext = React.createContext({});

function Controls() {
  let ref = React.createRef();
  const [prediction, setPrediction] = useState();

  return (
    <ControlsContext.Provider value={{ ref, prediction, setPrediction }}>
      <Answer />
      <Buttons />
      <Response />
    </ControlsContext.Provider>
  );
}

function Buttons() {
  const { ref, setPrediction } = useContext(ControlsContext);
  const { model } = useContext(GameContext);

  return (
    <div className="button-wrapper">
      <button
        onClick={() => {
          ref.current
            .getContext("2d")
            .clearRect(0, 0, ref.current.width, ref.current.height);
        }}
      >
        Clear
      </button>
      <button
        onClick={() => {
          predict(ref.current, model).then(prediction => {
            setPrediction(prediction);
          });
        }}
      >
        Predict
      </button>
    </div>
  );
}

function Response() {
  const { prediction } = useContext(ControlsContext);

  useEffect(() => {
    if (typeof prediction === "undefined") {
      console.log("Draw something!");
    } else {
      console.log("Did you mean: " + prediction + "?");
    }
  });

  return (
    <div className="response">
      <h2>Did you mean?</h2>
    </div>
  );
}

export default Controls;
export { ControlsContext };
