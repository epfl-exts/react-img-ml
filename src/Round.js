import React, { useContext, useState } from "react";
import Answer from "./Answer.js";
import Controls from "./Controls.js";

const RoundContext = React.createContext({});

function Round() {
  const [question, answered, setAnswered] = useQuestion();
  let ref = React.createRef();

  return (
    <div className={answered ? "round-wrapper roll-out-left" : "round-wrapper"}>
      <RoundContext.Provider value={{ question, ref, setAnswered }}>
        <div className="flex-wrapper">
          <Question />
          <Answer />
        </div>
        <Controls />
      </RoundContext.Provider>
    </div>
  );
}

function Question() {
  const { question } = useContext(RoundContext);
  const [num1, num2] = question;
  return (
    <div className="question">
      <span>{num1}</span>
      <h3>x</h3>
      <span>{num2}</span>
      <h3>=</h3>
    </div>
  );
}

function useQuestion() {
  const question = useRandomNumbers();
  const [answered, setAnswered] = useState(false);
  return [question, answered, value => setAnswered(value)];
}

function useRandomNumbers() {
  const limit = 9;
  const num1 = getRandomNumber(limit);
  const num2 = getRandomNumber(limit, num1);
  return [num1, num2];
}

function getRandomNumber(limit, multiple = 0) {
  const randomNumber = Math.floor(Math.random() * limit);
  if (randomNumber * multiple < 10) {
    return randomNumber;
  }
  return getRandomNumber(limit, multiple);
}

export default Round;
export { RoundContext };
