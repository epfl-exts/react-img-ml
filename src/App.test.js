import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import Round, { Question } from "./Round";
import Answer from "./Answer.js";
import Controls from "./Controls.js";

// describe("<Round />", () => {
//   it("renders correctly", () => {
//     const tree = renderer.create(<Round />).toJSON();
//     expect(tree).toMatchSnapshot();
//   });
// });

describe("<Round />", () => {
  it("renders: Question, Answer and Controls", () => {
    const wrapper = mount(<Round />);
    expect(wrapper.contains(<Question />)).toEqual(true);
    expect(wrapper.contains(<Answer />)).toEqual(true);
    expect(wrapper.contains(<Controls />)).toEqual(true);
  });
});
