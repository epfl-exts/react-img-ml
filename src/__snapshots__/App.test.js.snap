// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Round /> renders correctly 1`] = `
<div
  className="round-wrapper"
>
  <div
    className="flex-wrapper"
  >
    <div
      className="question"
    >
      <span>
        8
      </span>
      <h3>
        x
      </h3>
      <span>
        0
      </span>
      <h3>
        =
      </h3>
    </div>
    <canvas
      className="answer"
      height={320}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
      onMouseUp={[Function]}
      width={320}
    />
  </div>
  <div
    className="controls"
  >
    <button
      onClick={[Function]}
    >
      Clear
    </button>
    <button
      onClick={[Function]}
    >
      Predict
    </button>
  </div>
</div>
`;
